core/src/test/java/io/confluent/kafka/schemaregistry/SASLClusterTestHarness.java

```diff
@@ -0,0 +1,182 @@
+  public static final String JAAS_CONF = "java.security.auth.login.config";
+  public static final String KRB5_CONF = "java.security.krb5.conf";
+  public static final String ZK_AUTH_PROVIDER = "zookeeper.authProvider.1";
+
+  private static MiniKdc kdc;
+  private static File kdcHome;
+  private static String krb5ConfPath;
+  private static File jaasFile;
+
+  private static final Logger log = LoggerFactory.getLogger(SASLClusterTestHarness.class);
+
+  public SASLClusterTestHarness() {
+    super(DEFAULT_NUM_BROKERS);
+  }
+
+  @Override
+  protected SecurityProtocol getSecurityProtocol() {
+    return SecurityProtocol.SASL_PLAINTEXT;
+  }
+
+  @BeforeClass
+  public static void setUpKdc() throws Exception {
+    destroySaslHelper();
+    File zkServerKeytab = File.createTempFile("zookeeper-", ".keytab");
+    File kafkaKeytab = File.createTempFile("kafka-", ".keytab");
+    File schemaRegistryKeytab = File.createTempFile("schema-registry-", ".keytab");
+
+    // build and write the JAAS file.
+    JaasTestUtils.JaasSection serverSection = createJaasSection(zkServerKeytab,
+            "zookeeper/localhost@EXAMPLE.COM", "Server", "zookeeper");
+    // NOTE: there is only one `Client` section in the Jaas configuraiton file. Both the internal embedded Kafka
+    // cluster and the schema registry share the same principal. This is required because within the same JVM (eg
+    // these tests) one cannot have two sections, each with its own ZooKeeper client SASL credentials.
+    JaasTestUtils.JaasSection clientSection = createJaasSection(kafkaKeytab,
+            "kafka/localhost@EXAMPLE.COM", "Client", "zookeeper");
+    JaasTestUtils.JaasSection kafkaServerSection = createJaasSection(kafkaKeytab,
+            "kafka/localhost@EXAMPLE.COM", "KafkaServer", "kafka");
+    JaasTestUtils.JaasSection kafkaClientSection = createJaasSection(schemaRegistryKeytab,
+            "schema-registry/localhost@EXAMPLE.COM", "KafkaClient", "kafka");
+    jaasFile = File.createTempFile("schema_registry_tests", "_jaas.conf");
+    PrintWriter out = new PrintWriter(jaasFile);
+    out.println(serverSection.toString());
+    out.println(clientSection.toString());
+    out.println(kafkaServerSection.toString());
+    out.println(kafkaClientSection.toString());
+    out.close();
+
+    Properties kdcProps = MiniKdc.createConfig();
+    kdcHome = Files.createTempDirectory("mini-kdc").toFile();
+    log.info("Using KDC home: " + kdcHome.getAbsolutePath());
+    kdc = new MiniKdc(kdcProps, kdcHome);
+    kdc.start();
+
+    krb5ConfPath = System.getProperty("java.security.krb5.conf");
+
+    createPrincipal(zkServerKeytab, "zookeeper/localhost");
+    createPrincipal(kafkaKeytab, "kafka/localhost");
+    createPrincipal(schemaRegistryKeytab, "schema-registry/localhost");
+
+    // destroy SASL settings so all tests don't have SASL enabled. SASL is re-enabled in a @Before method below.
+    destroySaslHelper();
```

Comment:
I agree. I would like the KDC to be created before each test, not before the class. But tests fail when the KDC is created before each test, and I couldn't figure out why after 1-2 days of debugging. The downside to `@BeforeClass`, too, is all tests run slower.

I'll spend some time trying to get tests to pass with `@Before` instead of `@BeforeClass`.